### PR TITLE
Disallow bad combination of arguments in keygen grind

### LINKED ISSUE #17119: keygen --no-outfile args do not work as expected
#### Problem
When using --no-outfile on solana-keygen, it simpy exits the program after the grind is complete. 

solana@:~/ $ solana-keygen grind --no-outfile --starts-with xyz:1
Searching with 32 threads for:
        1 pubkey that starts with 'xyz' and ends with ''
Searched 1000000 keypairs in 1s. 0 matches found.
Searched 2000000 keypairs in 3s. 0 matches found.

### PR DESCRIPTION
**Pull request by steviez**
#### Problem
`solana-keygen grind` does not currently print the key when run with `--no-outfile`; if `--use-mnemonic` is not also specified, the result of this is useless as the private key is lost.

#### Summary of Changes
When `--no-outfile` is specified, require `--use-mnemonic`.
```
$ solana-keygen grind --starts-with s:1 --ignore-case --no-bip39-passphrase              
Searching with 12 threads for:
	1 pubkey that starts with 's' and ends with ''
Wrote grind keypair to sd3cFQrcrUi1hygi7QcTQPaXWUhbBHyFKPN9y79BUNV.json

// Usage that is now explicitly disallowed
$ solana-keygen grind --starts-with s:1 --ignore-case --no-bip39-passphrase --no-outfile               
error: The following required arguments were not provided:
    --use-mnemonic

USAGE:
    solana-keygen grind --config <FILEPATH> --ignore-case --language <LANGUAGE> --no-outfile --no-bip39-passphrase --num-threads <NUMBER> --starts-with <PREFIX:COUNT>... --use-mnemonic --word-count <NUMBER>

For more information try --help

$ solana-keygen grind --starts-with s:1 --ignore-case --no-bip39-passphrase --no-outfile --use-mnemonic
Searching with 12 threads for:
	1 pubkey that starts with 's' and ends with ''
================================================================================
Found matching key SPvXwbGcnjLif2JuXVUbLoTkUwYbntjLZDvW5sD5Qwi

Save this seed phrase to recover your new keypair:
unhappy loud crouch banana caught witness find polar century monster finger copy
================================================================================
```

Fixes #17119

### CODE REVIEW

**Comment by steviez** [comment]
Saw this issue and looked like low hanging fruit / a reason to play around with and learn a bit about the tool. Only thing I'm not sure of is the format of the print out. If we were writing to file, we'd dump the byte array, that is what I did here (and the option that was less work).

File: keygen/src/keygen.rs [hunk 1]
```diff
@@ -726,22 +728,25 @@ fn do_main(matches: &ArgMatches<'_>) -> Result<(), Box<dyn error::Error>> {
...
-                                    );
-                                }
+
+                                let outfile = {
+                                    if no_outfile {
+                                        STDOUT_OUTFILE_TOKEN.to_string()
+                                    } else {
+                                        format!("{}.json", keypair.pubkey())
+                                    }
+                                };
```
Comment (mvines) [inline keygen/src/keygen.rs:738]: nit:
```suggestion
                                let outfile =  if no_outfile {
                                        STDOUT_OUTFILE_TOKEN.to_string()
                                    } else {
                                        format!("{}.json", keypair.pubkey())
                                   };
```

File: keygen/src/keygen.rs [hunk 2]
```diff
@@ -726,22 +728,25 @@ fn do_main(matches: &ArgMatches<'_>) -> Result<(), Box<dyn error::Error>> {
...
+                                let _ = match output_keypair(&keypair, &outfile, "grind") {
+                                    Err(e) => println!("Unable to write {}: {}", outfile, e),
+                                    _ => {}
+                                };
```
Comment (mvines) [inline keygen/src/keygen.rs:742]: this is a little less verbose:
```suggestion
                                output_keypair(&keypair, &outfile, "grind")
                                    .unwrap_or_else(|e| eprintln!("Unable to write {}: {}", outfile, e));
```

**Review by mvines — APPROVED** [review]

File: keygen/src/keygen.rs (hunk 2 above)
Comment (steviez) [reply keygen/src/keygen.rs]: Good call on both of these, thanks!

**Comment by t-nelson** [comment]
The problem as described is expected behavior.  `--no-outfile` is intended to only output the seed phrase.  If the user wants the keypair text on stdout they should specify `-o -`

**Comment by steviez** [comment]
> The problem as described is expected behavior. `--no-outfile` is intended to only output the seed phrase. If the user wants the keypair text on stdout they should specify `-o -`

The `grind` subcommand does not support `--outfile`; rather, it only supports `--no-outfile`.
```
solana-keygen-grind 
Grind for vanity keypairs

USAGE:
    solana-keygen grind [FLAGS] [OPTIONS]

FLAGS:
    -h, --help                   Prints help information
        --ignore-case            Performs case insensitive matches
        --no-outfile             Only print a seed phrase and pubkey. Do not output a keypair file
        --no-bip39-passphrase    Do not prompt for a BIP39 passphrase
        --use-mnemonic           Generate using a mnemonic key phrase.  Expect a significant slowdown in this mode

OPTIONS:
    ...
```
Additionally, `--use-mnemonic` is an optional flag. So, if someone runs the script as follows with `--no-outfile` specified but without `--use-mnemonic`, they will get a public key printed out, but the private key is gone (not output to screen or disk, no seed phrase):
```
$ solana-keygen grind --no-outfile
```

This seems like a case we'd either want to avoid by either:
- Printing the private key to the screen if `--no-outfile` is passed (what I implemented here)
- Preventing `--no-outfile` (or maybe warning) if `--use-mnemonic` is not specified

Thoughts?

**Comment by CriesofCarrots** [comment]
>     * Preventing `--no-outfile` (or maybe warning) if `--use-mnemonic` is not specified

Personally, I like this approach

**Comment by t-nelson** [comment]
> > ```
> > * Preventing `--no-outfile` (or maybe warning) if `--use-mnemonic` is not specified
> > ```
> 
> Personally, I like this approach

Same.  Kill it with clap `Arg::conflicts()`

**Comment by steviez** [comment]
> > > ```
> > > * Preventing `--no-outfile` (or maybe warning) if `--use-mnemonic` is not specified
> > > ```
> > 
> > 
> > Personally, I like this approach
> 
> Same. Kill it with clap `Arg::conflicts()`

That seems reasonable to me - and thanks for chiming in @t-nelson & @CriesofCarrots; I saw you were both involved in the previous PR that added mnemonic support to grind, but was worried about being "trigger-happy" on adding too many reviewers ... next time I won't be because you found it anyways 😄

File: keygen/src/keygen.rs [hunk 3]
```diff
@@ -34,6 +34,8 @@ use std::{
 };
 
 const NO_PASSPHRASE: &str = "";
+// Sentinel value used to write to screen instead of file
+const STDOUT_OUTFILE_TOKEN: &str = "-";
```
Comment (t-nelson) [inline keygen/src/keygen.rs:38]: Would you mind moving this to `clap-utils` and replacing usage there as well? :pray: 

https://github.com/solana-labs/solana/blob/d1db5448b9c3c4f00301163196394e6bb9914a1d/clap-utils/src/keypair.rs#L238

**Review by t-nelson — DISMISSED** [review]
lgtm with the stdout constant moved to `clap-utils`.  Thanks for cleaning this up!

**Review by CriesofCarrots — DISMISSED** [review]
r+ Trent's suggestion of sharing the `const` with clap-utils. Thanks!

**Pull request merged**